### ISSUE TITLE
specialize types for different use cases

### ISSUE THREAD

**Issue by jadamcrain** [stepfunc/oo_bindgen#60]
Currently, we have a `Type` class that we reuse almost everywhere.

This causes a lot of complexity in code generators because not all of the use cases make sense or aren't needed.

Sometimes, conversion really only makes sense in one direction (e.g. bindings -> Rust) but not the other way around.

I think that we can agree that there are some types that always make sense bidirectionally and in almost any context, namely the primitive types.

```
enum PrimitiveType {
    Bool,
    Uint8,
    Sint8,
    Uint16,
    Sint16,
    Uint32,
    Sint32,
    Uint64,
    Sint64,
    Float,
    Double,
    String,
    Duration(DurationMapping),
    Enum(NativeEnumHandle),
}
```

I include `Duration` and `Enum` because they always has a straight-forward bi-directional mapping.

One easy target for reducing complexity is to simplify what can be in a `Struct`. I think it shouldn't be anything other than:

```
enum StructFieldType {
   Primitive(PrimitiveType),
   InnerStruct(NativeStructHandle)
}
```

Structs should just be used as **data**.  Is there anywhere that we use them otherwise? We don't need these things in structs (I think):

* Classes/ClassRefs
* Interfaces
* Iterators
* Collections
* StructRef

I'll extend this thread tomorrow afternoon with some more thoughts

**Comment by emgre** [comment]
I went through the DNP3 schema and the only exception I found is the OctetString struct used in the `ReadHandler`. It has a uint16 index value and an iterator of bytes.

We could possibly reuse the `StructElementType` that I introduced when I implemented the default values of structs. I had a `impl From<Type> for StructElementType` that made a `StructElementType` without a default value, but we could change it to have a `impl From<Primitive> for StructElementType` instead.
